Update the table-driven framework to support x86 imm-intrinsics

fiigii opened the pull request:
This PR updates the table-driven framework to support x86 imm-intrinsics. After this update, implementing imm-intrinsics just needs one-line definition in `hwintrinsiclistxarch.h`.

This PR implements certain imm-intrinsics which have special semanctics
1. `Avx.Compare\CompareScalar` just accepts imm-value in 0-31. Out of range argument will trigger `ArgumentOutOfRangeException`.
2. `Avx2\Sse2.ShiftLeftLogical\ShiftRightLogical\ShiftRightArithmetic` never throw any exception
    - that will be converted to `shift* ymm, ymm, xmm` (or `shift* xmm, xmm`)form when the imm-arg is not a constant. For example, `Avx2.ShiftLeftLogical(v, i)` would generate `vmovd   xmm1, edi;  vpsllw  ymm0, ymm0, xmm1` instead of the jumptable fallback.
    - <del>`ShiftLeftLogical\ShiftRightLogical` will be converted to `SetZeroVector256\128` on out of range imm-arg.</del>
    - <del>`ShiftRightArithmetic`'s imm-arg will be normalized to the maximum valid value on out of range imm-arg.</del>
3. <del>`Avx2\Sse2.ShiftLeftLogical128BitLane` will be converted to `SetZeroVector256\128` on out of range imm-arg.</del>

fiigii commented:
I will submit test cases later.

@CarolEidt @tannergooding @mikedn PTAL

4creators commented on src/jit/emitxarch.cpp:
```diff
@@ -170,6 +170,14 @@ bool emitter::IsDstDstSrcAVXInstruction(instruction ins)
         case INS_psubusb:
         case INS_psubusw:
         case INS_psubw:
+        case INS_psllw:
```
nit: it would help if we would keep alphabetical order of instructions

fiigii replied:
Thanks, will fix.

mikedn commented on src/jit/emitxarch.cpp:
```diff
@@ -5324,6 +5334,41 @@ void emitter::emitIns_SIMD_R_R_R(instruction ins, emitAttr attr, regNumber reg,
 
+void emitter::emitIns_SIMD_R_R_I(instruction ins, emitAttr attr, regNumber reg, regNumber reg1, int ival)
+{
+    bool isSSEShift = false;
```
Maybe this should be in a separate "isSseShift" function?

fiigii replied:
Yes, I will refactor here.

mikedn commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -131,7 +148,26 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                {
+                    if (op3->IsCnsIntOrI())
+                    {
+
```
Extra empty line

fiigii replied:
Thanks!

mikedn commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -93,17 +93,34 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                {
+                    if (op2->IsCnsIntOrI())
+                    {
+                        ssize_t ival = op2->AsIntConCommon()->IconValue();
```
Since you have used `IsCnsIntOrI` using `AsIntCon` would make more sense. `AsIntConCommon` also accepts `GT_CNS_LNG` which is not needed here.

fiigii replied:
Good point, thank you!

mikedn commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -413,6 +449,83 @@ void CodeGen::genHWIntrinsic_R_R_RM_I(GenTreeHWIntrinsic* node, instruction ins)
+    assert(op2Reg != REG_NA);
+    NamedIntrinsic  intrinsicID    = node->gtHWIntrinsicId;
+    var_types       baseType       = node->gtSIMDBaseType;
+    emitAttr        attr           = (emitAttr)node->gtSIMDSize;
```
I don't think it makes a significant difference but in many places the `EA_ATTR` macro is used for this.

mikedn commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -413,6 +449,83 @@ void CodeGen::genHWIntrinsic_R_R_RM_I(GenTreeHWIntrinsic* node, instruction ins)
+
+    // Emit the jump table
+
+    JITDUMP("\n      J_M%03u_DS%02u LABEL   DWORD\n", Compiler::s_compMethodsCount, jmpTableBase);
```
I don't think this is really useful. It looks like it was copied from `genJumpTable` but then it didn't belong there either. This kind of dump messages, if really needed, should be in `emitBBTableDataGenBeg` and `emitDataGenData`.

Or better, the data section should be included in the disassembly produced by the JIT. But that's a different story.

mikedn commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -295,6 +304,163 @@ GenTree* Compiler::getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE
+// Return Value:
+//     the max imm-value of non-full-range IMM intrinsic
+//
+int Compiler::immUpperBoundOfHWIntrinsic(NamedIntrinsic intrinsic, var_types baseType)
```
I just don't understand why people insist in adding more stuff to `Compiler`...

fiigii replied:
Hmm.. these functions are needed by multiple passes (e.g., importer, codegen, lowering, etc). Is there other better approach to organize these functions?

mikedn replied:
One reasonable option seems to be for these functions to be static members of `HWIntrinsicInfo`. Haven't really checked if this is actually possible or not. Another option might be global functions, similar to `varType` functions for example. Yet another option would be a `HWIntrinsicTable` class that basically acts as an abstraction for `hwIntrinsicInfoArray`.

That said, I'm not sure if it's worth bothering with this ATM.

mikedn commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -295,6 +304,163 @@ GenTree* Compiler::getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE
+GenTree* Compiler::impOutOfRangeFallback(NamedIntrinsic intrinsic, var_types simdType, var_types baseType)
+{
+    assert((flagsOfHWIntrinsic(intrinsic) & HW_Flag_OutOfRangeFallbackIMM) != 0);
+    assert(simdType == TYP_SIMD32 || simdType == TYP_SIMD16);
```
I see that the formatter doesn't flag these but the usual practice is to parenthesize this kind of expressions: `assert((simdType == TYP_SIMD32) || (simdType == TYP_SIMD16));`. Happens in multiple places.

mikedn commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -444,16 +610,46 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
+            // - AND, the imm-intrinsic does not have out-of-range fallback
+
+            if (!mustExpand && lastOp->IsCnsIntOrI() &&
+                lastOp->AsIntConCommon()->IconValue() > immUpperBoundOfHWIntrinsic(intrinsic, baseType))
```
`AsIntCon`

tannergooding commented on src/jit/codegenlinear.h:
```diff
@@ -134,6 +134,8 @@ void genFMAIntrinsic(GenTreeHWIntrinsic* node);
 void genLZCNTIntrinsic(GenTreeHWIntrinsic* node);
 void genPCLMULQDQIntrinsic(GenTreeHWIntrinsic* node);
 void genPOPCNTIntrinsic(GenTreeHWIntrinsic* node);
+void genJumpTableFallback(
```
I think @sdmaclea had a good approach to this in one of his PRs.

Specifically, the codegen is essentially the same except for the actual emitted instruction and could be handled via a lambda.

tannergooding replied:
https://github.com/dotnet/coreclr/pull/16085/files#diff-ddd0dded7caa705d00fda562d726c37aR5180
and
https://github.com/dotnet/coreclr/pull/16085/files#diff-ddd0dded7caa705d00fda562d726c37aR5266

4creators commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -295,6 +304,163 @@ GenTree* Compiler::getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE
+        case NI_SSE2_ShiftRightLogical128BitLane:
+            impPopStack();
+            impPopStack();
+            return gtNewSimdHWIntrinsicNode(simdType, NI_SSE2_SetZeroVector128, baseType, genTypeSize(simdType));
```
This can be combined with AVX2 with if / else clause for choosing `NI_SSE2_SetZeroVector128` or `NI_AVX_SetZeroVector256` intrinsic parameter

fiigii commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

test OSX10.12 x64 Checked jitincompletehwintrinsic
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx2
test OSX10.12 x64 Checked jitx86hwintrinsicnosimd
test OSX10.12 x64 Checked jitnox86hwintrinsic

4creators commented on src/jit/hwintrinsiclistxarch.h:
```diff
@@ -144,6 +144,7 @@ HARDWARE_INTRINSIC(SSE2_CompareNotLessThan,                          "CompareNot
 HARDWARE_INTRINSIC(SSE2_CompareNotLessThanOrEqual,                   "CompareNotLessThanOrEqual",                        SSE2,        6,           16,          2,            {INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_cmppd},             HW_Category_SimpleSIMD,                        HW_Flag_NoFlag)
 HARDWARE_INTRINSIC(SSE2_CompareOrdered,                              "CompareOrdered",                                   SSE2,        7,           16,          2,            {INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_cmppd},             HW_Category_SimpleSIMD,                        HW_Flag_NoFlag)
 HARDWARE_INTRINSIC(SSE2_CompareUnordered,                            "CompareUnordered",                                 SSE2,        3,           16,          2,            {INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_cmppd},             HW_Category_SimpleSIMD,                        HW_Flag_NoFlag)
+HARDWARE_INTRINSIC(SSE2_ConvertScalarToVector128Int32,               "ConvertScalarToVector128Int32",                    SSE2,       -1,           16,           1,           {INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_mov_i2xmm, INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid},           HW_Category_SimpleSIMD,                        HW_Flag_NoFlag)
```
It is implemented in #16237 as well

fiigii replied:
@4creators No worries, I just use this intrinsic as an internal helper. Once you implement it, I will remove here.

fiigii commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -428,6 +474,63 @@ void CodeGen::genHWIntrinsic_R_R_RM_I(GenTreeHWIntrinsic* node, instruction ins)
+    {
+        emit->emitIns_R_I(INS_cmp, EA_4BYTE, nonConstImmReg, maxImmValue);
+        emitJumpKind jge = genJumpKindForOper(GT_GE, CK_UNSIGNED);
+        genJumpToThrowHlpBlk(jge, SCK_ARG_RNG_EXCPN);
```
The runtime intermittently gets "failed to find exception throw block". @CarolEidt @tannergooding @mikedn @jkotas  Do you know why?

```
19:06:15         CLRJIT! CHECK::Trigger + 0x30E (0x7354b789)
19:06:15         CLRJIT! CodeGen::genJumpToThrowHlpBlk + 0x13C (0x73435bd7)
19:06:15         CLRJIT! CodeGen::genJumpTableFallback<<lambda_8c168ac1d3dcdd43425f08b8cdfa80f0> > + 0xF7 (0x7353e539)
19:06:15         CLRJIT! CodeGen::genHWIntrinsic + 0x2AE (0x7353ec79)
19:06:15         CLRJIT! CodeGen::genCodeForTreeNode + 0x43D (0x73516b55)
19:06:15         CLRJIT! CodeGen::genCodeForBBlist + 0x80E (0x7343936e)
19:06:15         CLRJIT! CodeGen::genGenerateCode + 0x283 (0x73434f23)
19:06:15         CLRJIT! Compiler::compCompile + 0x72D (0x73440571)
19:06:15         CLRJIT! Compiler::compCompileHelper + 0x589 (0x73441767)
19:06:15         CLRJIT! Compiler::compCompile + 0x622 (0x73440c53)
19:06:15             File: d:\j\workspace\x86_checked_w---b1af1919\src\jit\codegencommon.cpp Line: 2826
19:06:15             Image: D:\j\workspace\x86_checked_w---b1af1919\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root\CoreRun.exe
```

```cpp
        {
            /* Find the helper-block which raises the exception. */

            Compiler::AddCodeDsc* add =
                compiler->fgFindExcptnTarget(codeKind, compiler->bbThrowIndex(compiler->compCurBB));
            PREFIX_ASSUME_MSG((add != nullptr), ("ERROR: failed to find exception throw block"));
            tgtBlk = add->acdDstBlk;
        }
```

CarolEidt replied:
Range checks must be exposed in the importer, so that they can be appropriately modeled in the control flow. Check out the places in simd.cpp where it generates `GT_SIMD_CHK`.

fiigii replied:
@CarolEidt Thanks for pointing out. IMO, I still do not like the current solution to non-const (that hardcodes a jump-table fallback in the compiler), which
- needs more IR to represent the range-check because the out-of-range constant argument will become a variable in the recursive function-body.
- the codegen always needs to consider keeping consistent semantics with the front-end. 

So I think the software switch-table fallback can significantly simplify the compiler implementation that only deals with constant imm-argument. 

cc @tannergooding

tannergooding replied:
I thought we determined that, we could also do this as:
```C#
result Intrinsic(op1, op2)
{
    RangeCheck();
    return Intrinsic(op1, op2);
}
```

Which only leaves the switch table up to the JIT.

tannergooding replied:
If we do it all in the JIT however, it still seems like a one time cost in implementing a helper function to do this, and it only needs to happen in the fallback case, which is already on its own codepath.

fiigii replied:
I think the software switch-table is not a problem now. In the future, if the compiler determines to optimize away, we can use an attribute to disable the optimization on these intrinsics.

tannergooding replied:
> In the future, if the compiler determines to optimize away the switch, we can use an attribute to disable the optimization on these intrinsics.

There are multiple different compilers, and some (such as the C# compilers) don't have an option to disable optimizations for a given method. This would only work for an AOT or JIT compiler that was performing the optimization, and only if it respected the flag.

I think it would be good to better understand the specific pushback here. We already discussed this at some length in at least one of the other threads and landed on one of these two approaches:
* do the range check in managed and emit the switch table in the JIT
* emit the range check and the switch table in the JIT

CarolEidt replied:
As far as I'm concerned, it is a non-starter to consider introducing a throw in codegen that has not been introduced into the IR earlier. In addition to the technical infeasibility, it would render it impossible to apply range-check optimizations to eliminate it.

I am mildly in favor of emitting both the range check and the switch table in the JIT, primarily because it doesn't rely on the IL implementation and the JIT to both (and separately) specify the appropriate range values.

fiigii commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

test OSX10.12 x64 Checked jitincompletehwintrinsic
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx2
test OSX10.12 x64 Checked jitx86hwintrinsicnosimd
test OSX10.12 x64 Checked jitnox86hwintrinsic

fiigii commented:
@sandreenko The new range-check IR failed on the recent stack-level-setter changes: https://github.com/dotnet/coreclr/pull/15597/files#diff-b03137b61850257620434c9d13ff67d1R2825

fiigii commented:
Rebased and resolved a conflict from stack-level-setter.

tannergooding commented on src/jit/codegenxarch.cpp:
```diff
@@ -3756,7 +3759,10 @@ void CodeGen::genCodeForCmpXchg(GenTreeCmpXchg* tree)
 void CodeGen::genRangeCheck(GenTree* oper)
 {
-#ifdef FEATURE_SIMD
+#if defined(FEATURE_HW_INTRINSICS) && defined(FEATURE_SIMD)
```
Shouldn't the standalone `GT_HWIntrinsic_CHK` case also be handled (at the very least, based on the `ifdefs` elsewhere)?

fiigii replied:
So far, all the imm-intrinsics are SIMD. Do you think it is necessary?

tannergooding replied:
I think that all `#ifdef`s concerning the HWIntrinsic Check should be consistent. So they either all need to be under `#if defined(FEATURE_HW_INTRINSICS) && defined(FEATURE_SIMD)` or they should all be under `#ifdef FEATURE_HW_INTRINSICS`. Having a mix of both (when there isn't two actual distinctions of where the code applies) just makes it confusing (and prone to breakage) later down the road.

fiigii replied:
Got it, I will add a standalone `#ifdef GT_HWIntrinsic_CHK`.

tannergooding commented on src/jit/emitxarch.cpp:
```diff
@@ -5466,6 +5479,43 @@ void emitter::emitIns_SIMD_R_R_R(instruction ins, emitAttr attr, regNumber reg,
     }
 }
 
+static bool isSSEShift(instruction ins)
```
Is `Sse` the appropriate term here, since it also looks to cover AVX instructions, etc?

fiigii replied:
These special encoding shift instructions were introduced by SSE2, and AVX just gives them the VEX-encoding version. There are some AVX/AVX2-only shift instructions that don't belong to this group.

tannergooding commented on src/jit/emitxarch.cpp:
```diff
@@ -5466,6 +5479,43 @@ void emitter::emitIns_SIMD_R_R_R(instruction ins, emitAttr attr, regNumber reg,
+
+void emitter::emitIns_SIMD_R_R_I(instruction ins, emitAttr attr, regNumber reg, regNumber reg1, int ival)
+{
+    // TODO-XARCH refactoring emitIns_R_R_I to handle SSE2/AVX2 shift as well as emitIns_R_I
```
What about this isn't working today?

fiigii replied:
SSE2 shift instructions require special encoding, we need to improve `emitIns_R_R_I` to handle https://github.com/dotnet/coreclr/blob/master/src/jit/emitxarch.cpp#L3688-L3700 as well.

tannergooding replied:
Could we log an actual bug, as well as this TODO?

fiigii replied:
Logged at https://github.com/dotnet/coreclr/issues/16543

tannergooding commented on src/jit/gentree.cpp:
```diff
@@ -4911,7 +4927,10 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
+#endif // FEATURE_SIMD
+#ifdef FEATURE_HW_INTRINSICS
+        case GT_HWIntrinsic_CHK:
+#endif                  // FEATURE_HW_INTRINSICS
```
Would a newline after the `#ifdef` preventing the spacing weirdness on this comment?

fiigii replied:
Good catch! Thanks.

tannergooding commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -428,6 +472,56 @@ void CodeGen::genHWIntrinsic_R_R_RM_I(GenTreeHWIntrinsic* node, instruction ins)
+                                   regNumber                 offsReg,
+                                   HWIntrinsicSwitchCaseBody emitSwCase)
+{
+    assert(nonConstImmReg != REG_NA);
```
A comment about how/where inappropriate ranges are handled would be useful.

fiigii replied:
Added the comment in the function header.

tannergooding commented on src/jit/hwintrinsiclistxarch.h:
```diff
@@ -209,6 +209,11 @@ HARDWARE_INTRINSIC(SSE2_PackSignedSaturate,                          "PackSigned
+HARDWARE_INTRINSIC(SSE2_ShiftLeftLogical,                            "ShiftLeftLogical",                                 SSE2,       -1,           16,           2,           {INS_invalid,   INS_invalid,   INS_psllw,     INS_psllw,     INS_pslld,     INS_pslld,     INS_psllq,     INS_psllq,     INS_invalid,   INS_invalid},           HW_Category_IMM,                               HW_Flag_MaybeIMM|H ...
+HARDWARE_INTRINSIC(SSE2_ShiftLeftLogical128BitLane,                  "ShiftLeftLogical128BitLane",                       SSE2,       -1,           16,           2,           {INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_invalid,   INS_invalid},           HW_Category_IMM,                               HW_Flag_FullRangeI ...
+HARDWARE_INTRINSIC(SSE2_ShiftRightArithmetic,                        "ShiftRightArithmetic",                             SSE2,       -1,           16,           2,           {INS_invalid,   INS_invalid,   INS_psraw,     INS_invalid,   INS_psrad,     INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid},           HW_Category_IMM,                               HW_Flag_MaybeIMM|H ...
+HARDWARE_INTRINSIC(SSE2_ShiftRightLogical,                           "ShiftRightLogical",                                SSE2,       -1,           16,           2,           {INS_invalid,   INS_invalid,   INS_psrlw,     INS_psrlw,     INS_psrld,     INS_psrld,     INS_psrlq,     INS_psrlq,     INS_invalid,   INS_invalid},           HW_Category_IMM,                               HW_Flag_MaybeIMM|H ...
```
What is the distinction between the `MaybeImm` and `NoJmpTableImm` flags?

fiigii replied:
`MaybeImm` means this intrinsic has "imm" and "non-imm" overloads both.

`NoJmpTableImm` means certain imm-intrinsic has "variable byte" version that can be used as the non-const fallback. For example, 
```csharp
public static Vector256<ulong> ShiftRightLogical(Vector256<ulong> value, Vector128<ulong> count)
public static Vector256<ulong> ShiftRightLogical(Vector256<ulong> value, byte count)
```

fiigii replied:
Yes, I updated my comment.

CarolEidt replied:
This is still not clear to me. All instances of one also include the other. Can you give an example that would not have both flags?

tannergooding commented on src/jit/hwintrinsiclistxarch.h:
```diff
@@ -209,6 +209,11 @@ HARDWARE_INTRINSIC(SSE2_PackSignedSaturate,                          "PackSigned
 HARDWARE_INTRINSIC(SSE2_SetZeroVector128,                            "SetZeroVector128",                                 SSE2,       -1,           16,          0,            {INS_pxor,      INS_pxor,      INS_pxor,      INS_pxor,      INS_pxor,      INS_pxor,      INS_pxor,      INS_pxor,      INS_invalid,   INS_xorpd},             HW_Category_Helper,                            HW_Flag_NoFlag)
 HARDWARE_INTRINSIC(SSE2_SumAbsoluteDifferences,                      "SumAbsoluteDifferences",                           SSE2,       -1,           16,          2,            {INS_invalid,   INS_psadbw,    INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid},           HW_Category_SimpleSIMD,                        HW_Flag_BaseTypeFr ...
+HARDWARE_INTRINSIC(SSE2_ShiftLeftLogical,                            "ShiftLeftLogical",                                 SSE2,       -1,           16,           2,           {INS_invalid,   INS_invalid,   INS_psllw,     INS_psllw,     INS_pslld,     INS_pslld,     INS_psllq,     INS_psllq,     INS_invalid,   INS_invalid},           HW_Category_IMM,                               HW_Flag_MaybeIMM|H ...
+HARDWARE_INTRINSIC(SSE2_ShiftLeftLogical128BitLane,                  "ShiftLeftLogical128BitLane",                       SSE2,       -1,           16,           2,           {INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_invalid,   INS_invalid},           HW_Category_IMM,                               HW_Flag_FullRangeI ...
```
nit: Table alignment is off on some of the numbers

tannergooding commented on src/jit/hwintrinsiclistxarch.h:
```diff
@@ -297,11 +305,19 @@ HARDWARE_INTRINSIC(AVX2_HorizontalSubtractSaturate,                  "Horizontal
 HARDWARE_INTRINSIC(AVX2_Or,                                          "Or",                                               AVX2,       -1,           32,           2,           {INS_por,       INS_por,       INS_por,       INS_por,       INS_por,       INS_por,       INS_por,       INS_por,       INS_invalid,   INS_invalid},           HW_Category_SimpleSIMD,                        HW_Flag_Commutativ ...
+HARDWARE_INTRINSIC(AVX2_ShiftLeftLogical,                            "ShiftLeftLogical",                                 AVX2,       -1,           32,           2,           {INS_invalid,   INS_invalid,   INS_psllw,     INS_psllw,     INS_pslld,     INS_pslld,     INS_psllq,     INS_psllq,     INS_invalid,   INS_invalid},           HW_Category_IMM,                               HW_Flag_MaybeIMM|H ...
+HARDWARE_INTRINSIC(AVX2_ShiftLeftLogical128BitLane,                  "ShiftLeftLogical128BitLane",                       AVX2,       -1,           32,           2,           {INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_pslldq,    INS_invalid,   INS_invalid},           HW_Category_IMM,                               HW_Flag_FullRangeI ...
+HARDWARE_INTRINSIC(AVX2_ShiftLeftLogicalVariable,                    "ShiftLeftLogicalVariable",                         AVX2,       -1,           -1,           2,           {INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_vpsllvd,   INS_vpsllvd,   INS_vpsllvq,   INS_vpsllvq,   INS_invalid,   INS_invalid},           HW_Category_SimpleSIMD,                        HW_Flag_UnfixedSIM ...
```
What does `UnfixedSIMDSize` mean?

fiigii replied:
Some AVX/AVX2/FMA intrinsic operate over `Vector128/256` both.

tannergooding commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -185,15 +185,24 @@ int Compiler::ivalOfHWIntrinsic(NamedIntrinsic intrinsic)
     assert(intrinsic > NI_HW_INTRINSIC_START && intrinsic < NI_HW_INTRINSIC_END);
-    assert((hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].flags & HW_Flag_UnfixedSIMDSize) == 0);
-    return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].simdSize;
+    if ((Compiler::flagsOfHWIntrinsic(intrinsic) & HW_Flag_UnfixedSIMDSize) != 0)
```
Would the following be more readable?
```
simdSize = hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].simdSize;

if ((Compiler::flagsOfHWIntrinsic(intrinsic) & HW_Flag_UnfixedSIMDSize) == 0)
{
    return simdSize;
}

unsigned int sizeBytes = getSIMDTypeSizeInBytes(sig->retTypeSigClass);

assert(SimdSize == -1);
assert(sizeBytes != 0;

return sizeBytes
```

fiigii replied:
Looks good, will change.

tannergooding commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -295,6 +304,130 @@ GenTree* Compiler::getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE
+{
+    assert(categoryOfHWIntrinsic(intrinsic) == HW_Category_IMM);
+    HWIntrinsicFlag flags = flagsOfHWIntrinsic(intrinsic);
+    if ((flags & HW_Flag_FullRangeIMM) != 0)
```
Why not:

```
switch (intrinsic)
{
    case NI_AVX_Compare:
    case NI_AVX_CompareScalar:
        return 31;

    default:
        assert((flags & HW_Flag_FullRangeIMM) != 0);
        return 255;
}
```

fiigii replied:
Will change.

tannergooding commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -295,6 +304,130 @@ GenTree* Compiler::getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE
+}
+
+//------------------------------------------------------------------------
+// impNonConstFallback: convert certain SSE2/AVX2 shift intrinsic to its semantic alternative when the imm-arg is
```
This seems ok/like a good direction to me. But want to make sure that this is ok/inline with the overall feature design, @CarolEidt?

CarolEidt replied:
It seems reasonable to me, but there needs to be more clarity with respect to `HW_Category_IMM`, `HW_Flag_MaybeIMM` and `HW_Flag_NoJmpTableIMM`. In particular, the comments say that `HW_Category_IMM` intrinsics *require* immediate value, but there are entries in the table that have that set, as well as `HW_Flag_MaybeIMM`.

tannergooding commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -537,6 +694,8 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
                     strip(info.compCompHnd->getArgType(sig, info.compCompHnd->getArgNext(argList), &argClass)));
                 op2 = getArgForHWIntrinsic(argType, argClass);
 
+                op2 = addRangeCheck(intrinsic, op2, mustExpand);
```
Is this getting emitted even for the full range intrinsics?

fiigii replied:
I will change here to handle non-full-range intrinsics.

fiigii replied:
No, I meant I will change it in this PR 😄

fiigii replied:
Fixed in the function-body of `addRangeCheck`.

mikedn replied:
> Fixed in the function-body of addRangeCheck.

I'd suggest renaming to `addRangeCheckIfNeeded`.

fiigii replied:
> I'd suggest renaming to addRangeCheckIfNeeded.

Good point! will change.

tannergooding commented on src/jit/lowerxarch.cpp:
```diff
@@ -2409,14 +2409,29 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
+        }
 
-        if (op3->IsCnsIntOrI())
+        if (Compiler::isImmHWIntrinsic(intrinsicID, lastOp))
```
A comment explaining why we do this check in addition to the category check above would be useful

4creators commented:
@fiigii When do you expect this PR could be finished? This work is blocking me in #15777 which would complete implementation of SSE2 ISA.

fiigii commented:
@4creators I think it can be finished by the end of today, then wait for reviews.

4creators commented:
@fiigii Thanks, these are good news.

CarolEidt commented on src/jit/gtlist.h:
```diff
@@ -97,6 +97,10 @@ GTNODE(SIMD_CHK         , GenTreeBoundsChk   ,0,GTK_SPECIAL|GTK_NOVALUE)// Compa
 #endif // FEATURE_SIMD
 
+#ifdef FEATURE_HW_INTRINSICS
+GTNODE(HWIntrinsic_CHK  , GenTreeBoundsChk   ,0,GTK_SPECIAL|GTK_NOVALUE)
```
All the node names are all-caps. This shouldn't depart from that convention. I would suggest HW_INTRINSIC_CHK

tannergooding replied:
@CarolEidt, I think this also deviated with `GenTreeHWIntrinsic`. Should a bug be logged to fix that as well?

CarolEidt replied:
Ah yes, I didn't even realize that. GT_HWIntrinsic should probably be GT_HW_INTRINSIC. Thanks

CarolEidt commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -428,6 +472,56 @@ void CodeGen::genHWIntrinsic_R_R_RM_I(GenTreeHWIntrinsic* node, instruction ins)
 }
 
+template <typename HWIntrinsicSwitchCaseBody>
+void CodeGen::genJumpTableFallback(regNumber                 nonConstImmReg,
```
This method needs a function header comment.

fiigii commented:
> This is still not clear to me. All instances of one also include the other. Can you give an example that would not have both flags?

@CarolEidt The new-style intrinsic is based on function names rather than signatures,  and some intrinsics have "imm" and simple-SIMD overloads both. For example
```csharp
public static Vector256<sbyte> Shuffle(Vector256<sbyte> value, Vector256<sbyte> mask);
public static Vector256<int> Shuffle(Vector256<int> value, byte control) ;
```
so a category `HW_Category_IMM` is not enough to indicate a intrinsic requiring `imm8` or not (the above intrinsic should have flag `HW_Flag_MaybeIMM`). Consequently, we have to use the function `bool Compiler::isImmHWIntrinsic(NamedIntrinsic intrinsic, GenTree* lastOp)`.

In this kind of imm-intrinsic (`HW_Flag_MaybeIMM`), certain intrinsics are special, which have encoding that can carry the `imm8` operand in XMM registers, for example
```csharp
public static Vector256<ulong> ShiftRightLogical(Vector256<ulong> value, byte count);
public static Vector256<ulong> ShiftRightLogical(Vector256<ulong> value, Vector128<ulong> count);
                                                                          ^^^^^^
```
So, this intrinsic can have flag `HW_Flag_NoJmpTableIMM` to use XMM version as the non-constant fallback.

fiigii commented:
Addressed all the feedback, I think this PR is ready for final review.
JIT change is in the latest commit, hope it readable. 

@CarolEidt @tannergooding PTAL

fiigii commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

test OSX10.12 x64 Checked jitincompletehwintrinsic
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx2
test OSX10.12 x64 Checked jitx86hwintrinsicnosimd
test OSX10.12 x64 Checked jitnox86hwintrinsic

tannergooding commented on src/jit/codegenlinear.h:
```diff
@@ -135,6 +132,11 @@ void genFMAIntrinsic(GenTreeHWIntrinsic* node);
 void genPCLMULQDQIntrinsic(GenTreeHWIntrinsic* node);
 void genPOPCNTIntrinsic(GenTreeHWIntrinsic* node);
+template <typename HWIntrinsicSwitchCaseBody>
+void genJumpTableFallback(regNumber                 nonConstImmReg,
```
Should this be called `genHWIntrinsic_JumpTableFallback` (since it is specific to HWIntrinsic codegen)?

fiigii replied:
>  it is specific to HWIntrinsic 

Both okay to me. We do have other "JumpTableFallback" now.

tannergooding replied:
Yes, but we don't want future code to think it is safe to use as general purpose (if it isn't, and if it is, the typename should be changed).

fiigii replied:
Okay, will change.

fiigii commented on tests/src/JIT/HardwareIntrinsics/X86/Shared/GenerateTests.csx:
```diff
@@ -309,6 +356,53 @@ private static readonly (string templateFileName, string[] templateData)[] Avx2I
+    ("ImmUnOpTest.template",    new string[] { "Avx2",  "Avx",    "ShiftRightLogical128BitLane",          "Vector256",  "Int32",     "Vector256",  "Int32",           "1",         "32",      "(int)8",                                                  "result[0] != 134217728",                                                           "(i == 3 || i == 7 ? result[i] != 0 : result[i] != 134217728)"}),
+    ("ImmUnOpTest.template",    new string[] { "Avx2",  "Avx",    "ShiftRightLogical128BitLane",          "Vector256",  "UInt32",    "Vector256",  "UInt32",          "1",         "32",      "(uint)8",                                                 "result[0] != 134217728",                                                           "(i == 3 || i == 7 ? result[i] != 0 : result[i] != 134217728)"}),
+    ("ImmUnOpTest.template",    new string[] { "Avx2",  "Avx",    "ShiftRightLogical128BitLane",          "Vector256",  "Int64",     "Vector256",  "Int64",           "1",         "32",      "(long)8",                                                 "result[0] != 576460752303423488L",                                                 "(i == 2 ? result[i] != 576460752303423488L : result[i] != 0)"}),
+    ("ImmUnOpTest.template",    new string[] { "Avx2",  "Avx",    "ShiftRightLogical128BitLane",          "Vector256",  "UInt64",    "Vector256",  "UInt64",          "1",         "32",      "(ulong)8",                                                "result[0] != 576460752303423488UL",                                                "(i == 2 ? result[i] != 576460752303423488UL : result[i] != 0)"}),
```
Added a new template for imm-intrinsics that fixes imm-argument to simplify the verification.

tannergooding replied:
So this is a test which covers the case where `INS m128, imm8` gets compiled down to `INS m128, m128`, correct?

fiigii replied:
Yes, the reflection scenarios covered this codegen.

fiigii commented on src/jit/emitxarch.cpp:
```diff
@@ -10724,7 +10774,6 @@ BYTE* emitter::emitOutputRI(BYTE* dst, instrDesc* id)
         // Handle SSE2 instructions of the form "opcode reg, immed8"
 
         assert(id->idGCref() == GCT_NONE);
-        assert(valInByte);
```
Removed an assertion. This did not work with the jump-table fallback "opcode reg, (128-255)".

tannergooding replied:
That seems strange, I would expect that 128-255 would qualify as "fits in byte"

tannergooding replied:
@CarolEidt, do you have any context as to why `fitsInByte` is really `fitsInSByte`?

fiigii replied:
Seems we cannot directly remove this assertion. On 32-bit platform (x86 build), an strange assertion failure on `vpslldq  xmm0, 128`
```
Assert failure(PID 96 [0x00000060], Thread: 10976 [0x2ae0]): Assertion failed 'sz == _idCodeSize' in 'System.Runtime.Intrinsics.X86.Sse2:ShiftLeftLogical128BitLane(struct,ubyte):struct' (IL size 8)

    File: d:\workspace\coreclr-imm\src\jit\emit.h Line: 997
    Image: D:\workspace\coreclr-imm\bin\tests\Windows_NT.x86.Debug\Tests\Core_Root\CoreRun.exe
```

tannergooding replied:
I realized this is just an issue with casting.

We are taking an `unsigned int8` and converting it to a `signed native int`, which causes zero extension.

We need to first cast to `signed int8` so that we wind up with a signed extension instead.

tannergooding commented on src/jit/compiler.h:
```diff
@@ -3114,12 +3114,16 @@ class Compiler
     GenTree* getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE argClass);
-    GenTreeArgList* buildArgList(CORINFO_SIG_INFO* sig);
+    static int immUpperBoundOfHWIntrinsic(NamedIntrinsic intrinsic);
+    GenTree* impNonConstFallback(NamedIntrinsic intrinsic, var_types simdType, var_types baseType);
```
nit: It might be good to reorganize this code later. We have table lookup methods mixed in with importer methods.

tannergooding commented on src/jit/emitxarch.cpp:
```diff
@@ -10724,7 +10774,6 @@ BYTE* emitter::emitOutputRI(BYTE* dst, instrDesc* id)
         // Handle SSE2 instructions of the form "opcode reg, immed8"
 
         assert(id->idGCref() == GCT_NONE);
-        assert(valInByte);
```
Why did this assert get removed?

fiigii replied:
Because `valInByte` would be true only if imm-atgument is in 0-127.

tannergooding replied:
Commented here: https://github.com/dotnet/coreclr/pull/16183#discussion_r170110807, we can continue the conversation there to prevent two threads 😄

tannergooding commented on src/jit/gtlist.h:
```diff
@@ -97,6 +97,10 @@ GTNODE(SIMD_CHK         , GenTreeBoundsChk   ,0,GTK_SPECIAL|GTK_NOVALUE)// Compa
 #endif // FEATURE_SIMD
 
+#ifdef FEATURE_HW_INTRINSICS
+GTNODE(HW_INTRINSIC_CHK  , GenTreeBoundsChk   ,0,GTK_SPECIAL|GTK_NOVALUE)
```
nit: We have a short explanatory comment after most of the other GTNODE definitions

fiigii replied:
Will do. Thank you!

fiigii replied:
Done.

tannergooding commented on src/jit/gtstructs.h:
```diff
@@ -85,7 +85,9 @@ GTSTRUCT_1(FptrVal     , GT_FTN_ADDR)
 GTSTRUCT_1(Index       , GT_INDEX)
 GTSTRUCT_1(IndexAddr   , GT_INDEX_ADDR)
-#ifdef FEATURE_SIMD
+#if defined(FEATURE_HW_INTRINSICS) && defined(FEATURE_SIMD)
```
We covered four cases (Just HWIntrinsic, Just SIMD, both HWIntrinsic and SIMD, and Neither) in the similar `#ifdef` above, we should do the same here.

fiigii replied:
Fixed.

tannergooding commented on src/jit/compiler.h:
```diff
@@ -3114,12 +3114,16 @@ class Compiler
     bool compSupportsHWIntrinsic(InstructionSet isa);
     bool isScalarISA(InstructionSet isa);
     static int ivalOfHWIntrinsic(NamedIntrinsic intrinsic);
+    unsigned simdSizeOfHWIntrinsic(NamedIntrinsic intrinsic, CORINFO_SIG_INFO* sig);
```
Can this one not be static anymore?

fiigii replied:
No, it is calling `getSIMDTypeSizeInBytes`.

tannergooding commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -445,16 +570,40 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
     }
-    else if (category == HW_Category_IMM)
+    // Avoid checking stacktop for 0-op intrinsics
+    if (sig->numArgs > 0 && isImmHWIntrinsic(intrinsic, impStackTop().val))
```
Why `if` instead of `else if`?

fiigii replied:
From here, each `if` is not exclusive to others. For example, an intrinsic could be true at `if (sig->numArgs > 0 && isImmHWIntrinsic(intrinsic, impStackTop().val))` and `if ((flags & (HW_Flag_OneTypeGeneric | HW_Flag_TwoTypeGeneric)) != 0)`.

tannergooding commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -445,16 +570,40 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
+                return nullptr;
+            }
+
+            // TODO-XARCH-CQ repleace imm-intrinsic by certain lower-latency intrinsic based-on imm-values.
```
nit: I think this is one of the optimizations we might not want to make since it violates the contract of emitting a particular instruction.

tannergooding replied:
A performance analyzer would be well-suited to calling this type of optimization out instead.

fiigii replied:
C++ compilers (e.g., Clang, ICC) is making this optimization.

tannergooding replied:
Yes, they also make other optimizations such as using `UNPCKHPS` instead of `SHUFPS` in some cases.

However, I believe we had determined that we should keep this simple and emit exactly what the user requested under the assumption that they had manually tuned the code. (CC. @CarolEidt, in case I misunderstood).

The exception to this was the fallback case where we:
* Emit a jump table that will ultimately call the exact instruction
or 
* Resolve it to the same instruction, but with a different encoding which did not require the immediate parameter (some shift instructions)

fiigii replied:
> user requested under the assumption that they had manually tuned the code. 

Actually, I still think we should enable some optimizations for HW intrinsics (e.g., CSE) in the future. HW intrinsic is not only used as "higher-level assembly" to optimize a few functions, users can write complex fully-vectorized programs.

4creators replied:
IMO this should be solved by compiler hints. Developers will know when they can allow optimizations to be on and where optimizations should be always off. However, I would defer implementing this after first intriniscs implementation is shipped.

CarolEidt replied:
I suspect that our strategy here will evolve over time, but I think we should lean toward a faithful instantiation of the requested intrinsics. Over time we might want to consider what transformations or optimizations might be supported, but that makes sense as a future item.

CarolEidt replied:
I would urge you to just remove this comment.

tannergooding commented on src/jit/lowerxarch.cpp:
```diff
@@ -2409,24 +2409,29 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
     }
-    else if (numArgs == 3)
+
+    if (Compiler::categoryOfHWIntrinsic(intrinsicID) == HW_Category_IMM)
```
Why `if` instead of `else if`?

tannergooding replied:
Nevermind, to handle the two-op case (such as shifting).

we might want to return, instead of break, on a couple of the cases above, since they are known to not have the second-operand be imm.

fiigii replied:
Because imm-intrinsics can have 2 operands.

fiigii commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

test OSX10.12 x64 Checked jitincompletehwintrinsic
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx2
test OSX10.12 x64 Checked jitx86hwintrinsicnosimd
test OSX10.12 x64 Checked jitnox86hwintrinsic

fiigii commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

fiigii commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

fiigii commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -99,17 +99,49 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                        {
+                            // TODO-XARCH-BUG: the emitter cannot work with psrldq/pslldq on i > 127.
+                            //                 Use a same-semantic alternative as the workaround
+                            emit->emitIns_SIMD_R_R_R(INS_xorps, simdSize, targetReg, targetReg, targetReg);
```
@CarolEidt @tannergooding The emitter cannot work with `psrldq/pslldq` on `imm8` > 127. So I made this workaround and logged this bug at https://github.com/dotnet/coreclr/issues/16543. 
Please see https://github.com/dotnet/coreclr/pull/16183#discussion_r170104457 for more contexts.

tannergooding replied:
That is, I presume the problem is that we are coverting a `unsigned int8` to a `signed native int`, which will zero extend (255 -> 255), we want it to sign-extend instead so we would need to convert `unsigned int8` to `signed int8` (255 -> -1) and then to `signed native int` so it will sign extend (-1 -> -1)

fiigii replied:
> Can we not convert to unsigned char and then to char in order to ensure that "must be signed" semantics work?

No, I tried. And other code relies on the current behavior. Additionally, this bug is not only related to “valInByte”, 32-bit build has more problems...

tannergooding replied:
Could you elaborate?

The `valInByte` check should succeed if you add a cast in the `emitIns_SIMD_*_I` methods, which should unblock the assert/etc.

What code is relying on the "current behavior" (I assume this means the zero-extension) if you do cast the value (immediate values for SIMD instructions should always be 8-bit)?

What other problems does the 32-bit build have?

mikedn replied:
> Can we not convert to unsigned char and then to char in order to ensure that "must be signed" semantics work?

No amount of casting will make a value greater than 127 fit in a signed imm8...

> TODO-XARCH-BUG

I'm not sure why is this a bug. Emitting `xorps` is fine. Emitting `pxor` is also an option if you're concerned about data bypass delays (though I'm not sure if that applies to xor and anyway it doesn't really matter). Emitting `psrldq reg, 16` is yet another option, if you really, really want to emit a shift instructions (though I have no idea why would you want that) or if you want to keep the code shorter and simpler (now that would be a good reason).

fiigii replied:
> if you really, really want to emit a shift instructions

Not really, actually I want to replace most of the shift instructions by XOR if possible. Please see  https://github.com/dotnet/coreclr/pull/16183/files#diff-46e84d907ca2183faf22c1345e4e04aeR588

mikedn replied:
Not sure if it's worth doing that all other the place, it seems unlikely that people will write `shift(x, large constant shift count)` on purpose. This is more likely to be useful in C++, where extensive use of template metaprogramming can lead to cases where a large shift count magically appears and people even expect it to work well.

It's IMO worth doing if it cuts down complexity in common JIT code paths. As in this case, why bother making `emitIns_R_I` more complex when you can just emit `xor`?

Last time when I tried to add a bunch of new special cases to the emitter I ended up with a 0.2% regression in instruction count and rolled back most of the changes...

fiigii replied:
>  This is more likely to be useful in C++, where extensive use of template metaprogramming can lead to cases where a large shift count magically appears and people even expect it to work well.

> Last time when I tried to add a bunch of new special cases to the emitter I ended up with a 0.2% regression in instruction count and rolled back most of the changes...

Thank you so much for teaching! Emitting `xor` for `psrldq/pslldq` looks like a good solution now.

mikedn replied:
> No amount of casting will make a value greater than 127 fit in a signed imm8...

Actually it should probably work. It's implementation behavior in C++ but then the emitter does it anyway so it the C++ compiler doesn't like it then we'd have bigger problems anyway.

Doesn't really matter as one way or another you need special casing to handle this, either here or in the emitter. I'd move this special casing to the newly added `emitIns_SIMD_R_R_I` that too has special casing for shifts.

fiigii commented:
@tannergooding Do you have any idea about the failures below? I cannot reproduce it on my local machine (Windows 10 + VS2015), and CI only failed on Windows.

```
Assertion failed 'compiler->m_pLowering->IsContainableHWIntrinsicOp(node, op2) || op2->IsRegOptional()' 
```

tannergooding commented:
Looks like you have a node which is marked as contained but for which `IsContainableHWIntrinsicOp` is not true

fiigii commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

fiigii commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

test OSX10.12 x64 Checked jitincompletehwintrinsic
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx2
test OSX10.12 x64 Checked jitx86hwintrinsicnosimd
test OSX10.12 x64 Checked jitnox86hwintrinsic

tannergooding commented:
(commenting here since GitHub has hidden the thread on this https://github.com/dotnet/coreclr/pull/16183#discussion_r170408497)

@fiigii, @mikedn, @CarolEidt we will likely have to fix the `valInByte` issue eventually and I would rather do it now. There are other SIMD instructions which take an immediate, and for which the values 128-255 are valid and which cannot be simplified to something else because they actually do something. These instructions will need to have `valInByte` report true.

It looks like the only reason they aren't failing for intrinsics like `Sse.Shuffle` is because they go down the `emitIns_R_R_I` and `emitIns_R_R_R_I` paths which don't do a `valInByte` check (where-as `emitIns_R_I` does the check and uses it in the emitted instruction size estimation logic).

tannergooding commented:
I also don't think we should be emitting `xor` in place of the user requested `shift` instruction. We've had several discussions about this before and AFAIR the general consensus was to emit exactly what the user requested, unless such an instruction was not encodable in which case we either throw an exception or emit a fallback which ultimately executes the exact instruction.

The reasoning for this has been that, for a given hardware intrinsic, we are providing a contract to emit a particular instruction and that we will not modify the inputs/outputs. We are not making any guarantees about the encoding that will be emitted for said instruction, however.

There were a couple reasons for this:
* This feature is largely targeted at performance driven scenarios.
  * We can generally assume the user has profiled this type of code and is aware of what they are doing
  * Certain transformations could actually hurt performance in some of these cases as developers may be targeting a very specific platform for which they are taking advantage of an underlying quirk/behavior. 
    * Not everyone is targeting Intel. Some people may be targeting other platforms (AMD, VIA, etc) and may be relying on different optimization guidelines
    * Even on a given architecture, the optimization guidelines can vary from generation to generation
* For the cases where the user may be doing something which could negatively impact codegen, providing an analyzer to flag these issues should be fairly trivial.
  * This could include things like not passing in a constant (which causes a fallback) or passing in a value which will always result in 0.
* Providing a contract that we will emit a particular instruction is simple and provides no guarantees about the way the a particular instruction will behave.
  * If we decide to emit an "equivalent" but different instruction (i.e. `SetZero` instead of `ShiftLeft 255`), we are now changing the contract to one where the intrinsic has a specific behavior.
    * This means we need to be testing the actual implementation of the intrinsic, rather than just testing that we are emitting the right instruction and passing the input/output unmodified.

mikedn commented:
> There are other SIMD instructions which take an immediate, and for which the values 128-255 are valid and which cannot be simplified to something else because they actually do something. These instructions will need to have valInByte report true.

mikedn commented on src/jit/gtlist.h:
```diff
@@ -97,6 +97,10 @@ GTNODE(SIMD_CHK         , GenTreeBoundsChk   ,0,GTK_SPECIAL|GTK_NOVALUE)// Compa
 #endif // FEATURE_SIMD
 
+#ifdef FEATURE_HW_INTRINSICS
+GTNODE(HW_INTRINSIC_CHK  , GenTreeBoundsChk   ,0,GTK_SPECIAL|GTK_NOVALUE)// Compare whether an imm8 argument is in the valid range, and throw ArgumentOutOfRangeException if not.
```
It appears that this new oper has the same functionality as the existing `SIMD_CHK`. If that's the case then the existing oper should be used instead of adding a new one.

fiigii replied:
> same functionality as the existing SIMD_CHK

But the semantic is different, `SIMD_CHK` is used to check array bounds.

mikedn replied:
No, `GT_ARR_BOUNDS_CHECK` does that. Well, sort of. They all do the same thing - check if an integer is `>= 0` and `< gtArrLen` where `gtArrLen` can be any arbitrary tree (at least in theory). The difference is that some JIT parts (anything that has to do with redundant range check elimination) pay attention only to `GT_ARR_BOUNDS_CHECK`.

CarolEidt replied:
> Even the existence of GT_SIMD_CHK is questionable, there's a good chance that GT_ARR_BOUNDS_CHECK could have been used for that as well.

It is possible that they could be merged, but at the time we added `GT_SIMD_CHK` there was code that relied on the certain expected code patterns for `GT_ARR_BOUND_CHECK`, and it was felt that it would be less risky to avoid overloading it. It may be reasonable to merge these but I don't see it as critical. It may make more sense to merge all three to a `GT_RANGE_CHECK` in future, if we determine that it is more efficient.

mikedn replied:
Yes, merging with GT_ARR_BOUNDS_CHECK is a separate issue. But we shouldn't be adding a 3rd range check node without a good reason.

CarolEidt replied:
I concede your point. These should probably be merged, though I am not 100% sure that is without risk, and I really don't want to hold this up for something that might lead to complications.

mikedn commented on src/jit/codegenxarch.cpp:
```diff
@@ -3756,11 +3759,16 @@ void CodeGen::genCodeForCmpXchg(GenTreeCmpXchg* tree)
 void CodeGen::genRangeCheck(GenTree* oper)
 {
-#ifdef FEATURE_SIMD
+#if defined(FEATURE_HW_INTRINSICS) && defined(FEATURE_SIMD)
```
This whole ifdef goo is really just `assert(oper->OperIsBoundsCheck());`

mikedn commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -544,6 +497,72 @@ void CodeGen::genHWIntrinsic_R_R_RM_I(GenTreeHWIntrinsic* node, instruction ins)
+//    nonConstImmReg - the register contains non-constant imm8 argument
+//    baseReg        - a register for the start of the switch table
+//    offsReg        - a register for the offset into the switch table
+//    emitSwCase     - the lambda to geenrate siwtch-case
```
generate switch case

mikedn commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -544,6 +497,72 @@ void CodeGen::genHWIntrinsic_R_R_RM_I(GenTreeHWIntrinsic* node, instruction ins)
+    assert(nonConstImmReg != REG_NA);
+    emitter* emit = getEmitter();
+
+    const int   maxByte = 256;
```
Why is this hardcoded to 256? Shouldn't be 32 for CMPSS instruction? And shifts too have much lower useful maximum. It may be fallback code but IMO it shouldn't be unnecessarily large.

fiigii replied:
Will change.

tannergooding replied:
@fiigii, are we differentiating between instructions where an out of range imm
* causes a fault
* is ignored
* is masked

I think we may want to consider how these are handled eventually as causing a fault is the only one we actually need to range check against. If an out of range immediate is ignored, it is simply a nop, and if the upper bits are just masked out, then the input is still valid, even if not "normal" (ex: only bits 0-3 are used and the upper bits are ignored/masked, so imm8=16 and imm8=0 do the same thing).

mikedn commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -544,6 +497,72 @@ void CodeGen::genHWIntrinsic_R_R_RM_I(GenTreeHWIntrinsic* node, instruction ins)
+    unsigned jmpTableOffs = 0;
+
+    // Emit the jump table
+    for (unsigned i = 0; i < maxByte; i++)
```
`maxByte` is `int` but `i` is unsigned. And the lambda parameter is `int` too. And then `emitIns_R_R_I` has an `int` parameter even though we only need a 8 bit value and `GenTreeIcon` uses `ssize_t` to represent the immediate value. Such a mix of integer types is dubious.

fiigii replied:
Will fix, thanks.

mikedn commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -1097,12 +1093,32 @@ void CodeGen::genAVXIntrinsic(GenTreeHWIntrinsic* node)
+        {
+            assert(op1 == nullptr);
+            assert(op2 == nullptr);
+            if (!compiler->compSupports(InstructionSet_AVX2) && varTypeIsIntegral(baseType))
```
I don't quite understand what's this `if` for and what the whole thing has to do with this PR. Can you clarify?

fiigii replied:
`SetZeroVector256` will generate `pxor` with integral base-typ, but `pxor` is a AVX2 instruction, so we generate `xorps` on AVX machines to avoid the data bypass penalty.

tannergooding replied:
@fiigii, given that AVX doesn't expose any integer operations anyways, might it be better to only expose the integer based `SetZeroVector256` on the AVX2 isa (I believe this is what we did for SSE/SSE2)?

tannergooding replied:
This would also ensure users are explicitly aware of the penalty they may receive.

mikedn replied:
SSE/SSE2 is somewhat simpler, and weirder at the same time. SSE ended up with a non generic `SetZeroVector128` that gives you a `Vector128<float>`. SSE2 has a generic `SetZeroVector128<T>()` that gives you exactly what? I would guess it won't give you a `Vector128<float>` because that's the job of SSE.

If you really want to have an exacting mapping between intrinsic methods and CPU instructions I'd avoid using generic methods. It ends up being just weird and not very helpful to the developer who's supposed to use these intrinsic.

mikedn replied:
Though as a developer I'd expect `new Vector128<float>()` or `default` to just work and emit the right code. Having to use something like `Sse2.SetZeroVector128<int>()` just to get a 0 is pretty obnoxius.

tannergooding replied:
> If you really want to have an exacting mapping between intrinsic methods and CPU instructions I'd avoid using generic methods.

We've actually done a pretty good job on that so far (at least I think). For the most part, we only use the generic methods when all 10 supported primitives (float, double, byte, sbyte, short, ushort, int, uint, long, and ulong) are supported for that intrinsic in that ISA.

We have been "exploding" the APIs in the cases where only a subset of the supported primtives are supported by that ISA.

I would guess that `SetZeroVector128` was just missed (this also doesn't use the naming convention we have been following, and should be called `SetZeroVector128Single`, etc).

CC. @eerhardt, @fiigii, @CarolEidt

tannergooding replied:
> Though as a developer I'd expect new Vector128<float>() or default to just work and emit the right code.

This should work (you should get zero), but it wouldn't use `xorps`, `xorpd`, or `pxor` unless the JIT was already configured to zero structs that way.

mikedn replied:
> but it wouldn't use xorps, xorpd, or pxor 

Well, it should. I suspect at the moment it will always generate `xorps` which is good enough.

mikedn replied:
> We have been "exploding" the APIs in the cases where only a subset of the supported primtives are supported by that ISA.

fiigii replied:
> I would guess that SetZeroVector128 was just missed (this also doesn't use the naming convention we have been following, and should be called SetZeroVector128Single, etc).

Ah, I think we should move `Sse2.SetZeroVector128<T>` to `Sse` to match the AVX API.

> given that AVX doesn't expose any integer operations anyways, might it be better to only expose the integer based SetZeroVector256 on the AVX2 isa (I believe this is what we did for SSE/SSE2)?

I do not think it is necessary, for we never promise the generated instructions of helper-intrinsics and the codgen may be changed due to ISA development. One example is `SetAllVector128/256<T>` can generate `broadcast` on AVX2 machines.

fiigii replied:
Logged at https://github.com/dotnet/coreclr/issues/16575.

mikedn commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -99,17 +99,47 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                else if (Compiler::isImmHWIntrinsic(intrinsicID, op2))
+                {
+                    auto emitSwCase = [&](int i) {
+                        if ((ins == INS_psrldq || ins == INS_pslldq) && i > 15)
```
This special case can/should be moved to `emitIns_SIMD_R_R_I`.

fiigii replied:
Will change.

mikedn commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -295,6 +301,125 @@ GenTree* Compiler::getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE
+            GenTree* op1 = impSIMDPopStack(simdType);
+            GenTree* tmpOp =
+                gtNewSimdHWIntrinsicNode(TYP_SIMD16, op2, NI_SSE2_ConvertScalarToVector128Int32, TYP_INT, 16);
+            return gtNewSimdHWIntrinsicNode(simdType, op1, tmpOp, intrinsic, baseType, genTypeSize(simdType));
```
So switch based fallback is handled in codegen but non-switch based fallback is handle in the importer? It's not immediately obvious why similar cases are handled differently.

fiigii replied:
Because an intrinsic that needs the jump-table fallback has to become a function call in the importer, but these shift intrinsics can be directly replaced.

mikedn replied:
> but these shift intrinsics can be directly replaced.

Yes, but does this replacement has to happen in the importer?

fiigii replied:
I think addressing it in the importer is simpler, which we do not need to care about register allocation or optimizations (in the future) for these intrinsics in the back-end.

mikedn commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -295,6 +301,125 @@ GenTree* Compiler::getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE
+    {
+        assert(!lastOp->IsCnsIntOrI());
+        GenTree* upperBoundNode = new (this, GT_CNS_INT) GenTreeIntCon(TYP_INT, immUpperBoundOfHWIntrinsic(intrinsic));
+        GenTree* index          = gtCloneExpr(lastOp);
```
I guess `lastOp` can't only be a `GT_LCL_VAR` - the intrinsic method argument, otherwise cloning an arbitrary tree could be problematic. If it's indded `GT_LCL_VAR` then an assert could be useful.

fiigii commented:
@mikedn Question. Do you know why we have this assertion? https://github.com/dotnet/coreclr/blob/516b973b52d16843c5c7f20c53bd246d62c55d4e/src/jit/emit.h#L991

mikedn commented:
> Question. Do you know why we have this assertion? 

`_idCodeSize` is a bit field of 4 bits. This kind of asserts are used to ensure that the value stored in the bitfield doesn't get accidentally truncated.

mikedn commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -295,6 +301,125 @@ GenTree* Compiler::getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE
+GenTree* Compiler::addRangeCheck(NamedIntrinsic intrinsic, GenTree* lastOp, bool mustExpand)
+{
+    assert(lastOp != nullptr);
+    if (mustExpand && ((flagsOfHWIntrinsic(intrinsic) & HW_Flag_FullRangeIMM) == 0) &&
```
It would be useful to add a comment explaining why "full range imm" intrinsics do not need a range check - because the imm parameter of the intrinsic method is a `byte`.

That BTW might be a bit unfortunate, at least for shift intrinsics. Both the C/C++ intrinsics and the C/C++/C# shift operators use `int` or `unsigned` for the shift count. Using `byte` can be considered as a rather arbitrary restriction but probably it doesn't matter as we expect people to supply constants.

fiigii replied:
> It would be useful to add a comment explaining why "full range imm" intrinsics do not need a range check

Will do.

fiigii commented:
@CarolEidt Could you please take a look when you get a chance? This PR is blocking certain subsequent work.

CarolEidt commented on src/jit/namedintrinsiclist.h:
```diff
@@ -79,7 +79,15 @@ enum HWIntrinsicFlag : unsigned int
+
+    // NoJmpTable IMM
+    // the imm intrinsic does not need jumptable fallback when it gets non-const argument
+    HW_Flag_NoJmpTableIMM = 0x2000,
```
I still don't understand why both of these are required. In what case would we require a jump table if there was a non-immediate form available, and conversely, in what case would there be a non-immediate form available but we still require a jump table? Are there cases where the values covered by an immediate are somehow broader or otherwise disjoint from the values covered by a non-immediate form? I note that the existing intrinsics you added always have neither or both of these flags set.

fiigii replied:
> This is still not clear to me. All instances of one also include the other. Can you give an example that would not have both flags?

@CarolEidt The new-style intrinsic is based on function names rather than signatures,  and some intrinsics have "imm" and simple-SIMD overloads both. For example
```csharp
public static Vector256<sbyte> Shuffle(Vector256<sbyte> value, Vector256<sbyte> mask);
public static Vector256<int> Shuffle(Vector256<int> value, byte control) ;
```
so a category `HW_Category_IMM` is not enough to indicate a intrinsic requiring `imm8` or not (the above intrinsic should have flag `HW_Flag_MaybeIMM`). Consequently, we have to use the function `bool Compiler::isImmHWIntrinsic(NamedIntrinsic intrinsic, GenTree* lastOp)`.

In this kind of imm-intrinsic (`HW_Flag_MaybeIMM`), certain intrinsics are special, which have encoding that can carry the `imm8` operand in XMM registers, for example
```csharp
public static Vector256<ulong> ShiftRightLogical(Vector256<ulong> value, byte count);
public static Vector256<ulong> ShiftRightLogical(Vector256<ulong> value, Vector128<ulong> count);
                                                                          ^^^^^^
```
So, this intrinsic can have flag `HW_Flag_NoJmpTableIMM` to use XMM version as the non-constant fallback.

> I note that the existing intrinsics you added always have neither or both of these flags set.

Consequently, `Avx2.Shuffle` will need `HW_Flag_MaybeIMM` but no `HW_Flag_NoJmpTableIMM`.

fiigii replied:
> So, this intrinsic can have flag HW_Flag_NoJmpTableIMM to use XMM version as the non-constant fallback (rather than a jump-table).

I believe that the ISA design (carry the imm8 operand in XMM registers) is intentional for this situation, and ARM also has similar instructions.

CarolEidt replied:
OK, thanks for the explanation.

CarolEidt commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -295,6 +336,135 @@ GenTree* Compiler::getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE
+        return false;
+    }
+
+    if ((flagsOfHWIntrinsic(intrinsic) & HW_Flag_MaybeIMM) != 0 && genActualType(lastOp->TypeGet()) != TYP_INT)
```
It is not obvious why this is correct - it needs comments. This relies on the observation that *all* the `HW_Flag_MaybeIMM` cases have *either* an immmediate integer argument, *or* a non-integer argument, but not a non-constant integer argument.
Perhaps your "Return Value" comment should say:

Return true iff the intrinsics is an imm-intrinsic overload.
Note that some intrinsics, with HW_Flag_MaybeIMM set, have both imm (integer immediate) and vector (i.e. non-TYP_INT) overloads.

fiigii replied:
Thank you! Will do.

tannergooding commented:
@fiigii, we should run the HWIntrinsic tests before merging.

NOTE: These jobs will currently fail due to an unrelated issue: https://github.com/dotnet/coreclr/issues/16566. I've held off on kicking these jobs myself until after the PR resolving that is merged: https://github.com/dotnet/coreclr/pull/16567

tannergooding commented:
@fiigii, I would also like to see bugs logged for the 2-3 pending issues before merging (they all correspond to `TODO` comments added in this PR)

tannergooding commented on src/jit/emitxarch.cpp:
```diff
@@ -5479,6 +5492,51 @@ void emitter::emitIns_SIMD_R_R_R(instruction ins, emitAttr attr, regNumber reg,
+            emitIns_R_R(INS_movaps, attr, reg, reg1);
+        }
+
+        if ((ins == INS_psrldq || ins == INS_pslldq) && ival > 15)
```
Why just `psrldq` and `pslldq` instead of `isSseShift`

fiigii replied:
Oops, will fix.

tannergooding commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -544,6 +488,75 @@ void CodeGen::genHWIntrinsic_R_R_RM_I(GenTreeHWIntrinsic* node, instruction ins)
+
+    const unsigned maxByte = (unsigned)Compiler::immUpperBoundOfHWIntrinsic(intrinsic) + 1;
+    assert(maxByte <= 256);
+    BasicBlock* jmpTable[256];
```
nit: `BasicBlock* jmpTable[maxByte]`?

fiigii replied:
`maxByte` is not a compile-time constant.

tannergooding replied:
right, missed that 😄

tannergooding commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -544,6 +488,75 @@ void CodeGen::genHWIntrinsic_R_R_RM_I(GenTreeHWIntrinsic* node, instruction ins)
+    for (unsigned i = 0; i < maxByte; i++)
+    {
+        jmpTable[i] = genCreateTempLabel();
+        JITDUMP("            DD      L_M%03u_BB%02u\n", Compiler::s_compMethodsCount, jmpTable[i]->bbNum);
```
nit: You removed `JITDUMP("\n      J_M%03u_DS%02u LABEL   DWORD\n", Compiler::s_compMethodsCount, jmpTableBase);` from above, but not this corresponding line.

fiigii replied:
@mikedn suggested removing that.

mikedn replied:
I think I suggested removing both :)

tannergooding replied:
Yeah, I was indicating that if one line is removed, the other should be as well.

fiigii replied:
@mikedn @tannergooding Got it! Thanks.

tannergooding commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -544,6 +488,75 @@ void CodeGen::genHWIntrinsic_R_R_RM_I(GenTreeHWIntrinsic* node, instruction ins)
+    for (unsigned i = 0; i < maxByte; i++)
+    {
+        genDefineTempLabel(jmpTable[i]);
+        emitSwCase((int)i);
```
Why not keep this as a `byte`?

fiigii replied:
The underlying functions all accept `int`, so it may not be necessary to introduce one more type to mess here...

mikedn replied:
But the loop variable is still unsigned :)

tannergooding commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

test OSX10.12 x64 Checked jitincompletehwintrinsic
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx2
test OSX10.12 x64 Checked jitx86hwintrinsicnosimd
test OSX10.12 x64 Checked jitnox86hwintrinsic

tannergooding commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

test OSX10.12 x64 Checked jitincompletehwintrinsic
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx2
test OSX10.12 x64 Checked jitx86hwintrinsicnosimd
test OSX10.12 x64 Checked jitnox86hwintrinsic

tannergooding commented on src/jit/codegenxarch.cpp:
```diff
@@ -3756,12 +3759,7 @@ void CodeGen::genCodeForCmpXchg(GenTreeCmpXchg* tree)
-    noway_assert(oper->OperGet() == GT_ARR_BOUNDS_CHECK);
-#endif // !FEATURE_SIMD
-
+    assert(oper->OperIsBoundsCheck());
```
Should this be kept as a `noway_assert`?

fiigii replied:
Not sure, but will fix.

mikedn replied:
No, `noway_assert` is pointless here. All other code gen functions use `assert`.

tannergooding commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -445,16 +616,36 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
-        if (!lastOp->IsCnsIntOrI() && !mustExpand)
+        // The imm-HWintrinsics that do not accept all imm8 values may throw exception or transform to other
+        // intrinsics when the imm argument is not in the valid range
+        if ((flags & HW_Flag_FullRangeIMM) == 0)
```
Can you log a bug tracking us adding a helper for checking if a flag is set or unset?

fiigii replied:
Will do, and this comment should be updated, thanks.

fiigii commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -544,6 +490,74 @@ void CodeGen::genHWIntrinsic_R_R_RM_I(GenTreeHWIntrinsic* node, instruction ins)
+    for (unsigned i = 0; i < maxByte; i++)
+    {
+        genDefineTempLabel(jmpTable[i]);
+        emitSwCase(i);
```
@tannergooding @mikedn Changed `emitSwCase` to accept `unsigned`, so that this function only uses `unsigned` variables.

tannergooding commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

test OSX10.12 x64 Checked jitincompletehwintrinsic
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx2
test OSX10.12 x64 Checked jitx86hwintrinsicnosimd
test OSX10.12 x64 Checked jitnox86hwintrinsic

fiigii commented:
@CarolEidt @tannergooding @mikedn Thank you so much for your reviwe. All the feedback should already get fixed. Logged the remaining works at https://github.com/dotnet/coreclr/issues/16594 https://github.com/dotnet/coreclr/issues/16575 https://github.com/dotnet/coreclr/issues/16543 https://github.com/dotnet/coreclr/issues/16568

tannergooding commented:
@fiigii thanks for responding to the feedback so far.

I'll get this merged after all the tests complete (and pass).

CarolEidt commented:
Thanks for the changes & responses. I'm happy with where we are on this.

fiigii commented:
@4creators This PR is going to be merged. Would you like to complete the SSE2 intrinsic implementation? As I know, the code freeze date of .NET Core 2.1 may be coming soon, we probably prefer to enable all the SSE/SSE2/SSE3/SSSE3/SSE4.1/AVX intrinsics in this version of runtime. 
Do you have any plan of SSE2 intrinsic? If you have any problem that cannot implement all of them in a short time, please just let me know, we can cooperate to speed up the process.

4creators commented:
@fiigii Thanks for information. I will try to implement remaining SSE2 intrinsics asap. What is a plan on AVX2? Do we intend to finish them for 2.1 release?

fiigii commented:
> What is a plan on AVX2? Do we intend to finish them for 2.1 release?

I will try my best to implement as many as AVX2 intrinsics (depends on the deadline).

tannergooding commented:
@4creators, if you could send me an e-mail (tagoo at microsoft dot com) it might be a bit easier to sync up on this without polluting the thread.

4creators commented:
> if you could send me an e-mail (tagoo at microsoft dot com)

@tannergooding  Just did it

fiigii commented:
OSX tests get build timed out. Other tests all passed.
